rpc: fix existing usage of drpcmetadata

suj-krishnan opened the pull request:
This change is a follow-up for the drpc metadata fixes done earlier and removes all existing uses of drpcmetadata in the code, replacing them with grpc metadata calls.

**Why this works:**
On the client side, drpc client framework code fixes take care of sending grpc metadata over the wire (Ref PR: https://github.com/cockroachdb/drpc/pull/22). On the server side, [this](https://github.com/cockroachdb/drpc/pull/26) PR introduces changes in the drpc server framework code to populate the metadata received over the wire as _grpc_ context metadata - this behaviour is configured on the server by instantiating the drpc server's manager with a newly added option (`GrpcMetadataCompatMode`). Thus, both drpc client and server can simply use grpc metadata calls to send and receive metadata. 

**Version upgrade scenarios w.r.t metadata handling**
Notably, this implementation will take care of mixed behaviour that is expected during a cluster upgrade to enable DRPC. More details below:
  
**Version V0 (current version)**
* DRPC is not enabled.  
* Codebase has a mix of grpc and a few drpc metadata calls \- drpc metadata calls are executed only if drpc is enabled.

**Version V1**
* Cockroach codebase has only **grpc** metadata calls  
* DRPC client side framework code reads both **drpc** and **grpc** metadata from outgoing context and sends it over the wire.
* DRPC server side framework code reads metadata from the wire and populates it into **grpc** metadata in the incoming context.

**Version V2**
* Cockroach codebase has only **drpc** metadata calls  
* DRPC client side framework code reads only **drpc** metadata from outgoing context and sends it over the wire.  
* DRPC server side framework code reads metadata from the wire and populates it into **drpc** metadata in the incoming context.

**Version V0 to V1 upgrade scenarios**  
- Deploy Version V1 on the cluster as a rolling upgrade  
- After all nodes are on the V1 binary, enable DRPC in the cluster settings.  
- Each node could read the cluster setting at a different time, and therefore start using DRPC at a different time, so we would have mixed behaviour during this time.

| Node1 | Node2 |
| :---- | :---- |
| **V0 (drpc disabled)** | **V1**  |
| Sends grpc request with grpc metadata | grpc server receives grpc metadata |
| **V1 (drpc disabled)** | **V0**  |
| Sends grpc request with grpc metadata | grpc server receives grpc metadata |
| **V1 (drpc enabled)** | **V1**  |
| 1\. Cockroach code populates grpc metadata. 2\. drpc client framework sends grpc metadata over the wire. | 3\. Metadata received over the wire by drpc server framework code (V1) and populated into grpc metadata in the context. 4\. Cockroach code reads metadata from incoming using grpc metadata calls. |
| **V1 (drpc not enabled)** | **V1**  |
| Sends grpc request with grpc metadata | grpc server receives grpc metadata |

**Version V1 to V2 upgrade scenarios**   
Can be done only after all nodes communicate only using drpc.

| Node1 | Node2 |
| :---- | :---- |
| **V1** | **V2** |
| 1\. Cockroach code (V1) populates grpc metadata. 2\. drpc client framework (V1) sends grpc metadata over the wire. | 3\. Metadata received over the wire by drpc server framework code (V2) and populated into drpc metadata in the context. 4\. Cockroach code (V2) reads metadata from incoming context using drpc metadata calls. |
| **V2**  | **V1** |
| 1\. Cockroach code (V2) populates drpc metadata. 2\. drpc client framework (V2) sends it over the wire. | 3\. Metadata received over the wire by drpc server (V1) and populated into grpc metadata in the context. 4\. Cockroach code (V1) reads metadata from incoming context using grpc metadata calls. |

PR for drpc fixes: https://github.com/cockroachdb/drpc/pull/26

Epic: CRDB-48924
Fixes: cockroachdb#160757
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/159883)
<!-- Reviewable:end -->

shubhamdhama reviewed (commented):
I think I understood one aspect of incorrectly, which is apparently not the case but can be taken as a suggestion:

I thought only gRPC metadata will be accepted as a the incoming metadata and only gRPC metadata will be emitted as an outgoing metadata. The cockroach codebase shouldn't be using DRPC metadata layer at all before the completion of migration.

This will remove the problem of maintaining two metadata in the context like we are doing here. Only gRPC metadata is the source of truth. 

This mean, for the backport let's just keep the `grpcutil.Clear...`. Later, in a non-backport PR, DRPC layer won't put the metadata into the context using drpcmetadata.Add but should only be relying on grpcmetadata.NewOutgoingContext.

What do you think?

shubhamdhama commented:
If we want to include this change in 26.1 then I think we should merge something basis, that is, avoid changing the DRPC layer if possible.

shubhamdhama commented on pkg/rpc/drpc_test.go:
```diff
@@ -92,7 +93,11 @@ func TestGatewayRequestDRPCRecoveryInterceptor(t *testing.T) {
+			gwRequestKey: "test",
+		})
+		// Create a context with the gateway metadata
+		ctx := metadata.NewIncomingContext(context.Background(), md)
```
It should be using `MarkDRPCGatewayRequest`

shubhamdhama reviewed (commented):
LGTM!
I've left some comments. 

Please make the `GrpcMetadataCompatMode` option to be true by default in the framework, because we don't expect to use drpcmetadata anytime soon in CockroachDB.

suj-krishnan commented:
> LGTM! I've left some comments.
> 
> Please make the `GrpcMetadataCompatMode` option to be true by default in the framework, because we don't expect to use drpcmetadata anytime soon in CockroachDB.

The option is hardcoded to `true` in the cockroach server code, which is sort of the same thing. I prefer this since it makes the behaviour clear/explicit. Any particular reason you want to impose it at the framework?

shubhamdhama reviewed (commented):
LGTM! Thanks for answering my questions. I'll approve the PR once the below comment is addressed and personal fork is removed from the dependency.

suj-krishnan committed:
rpc: remove existing usage of drpcmetadata
This change is a follow-up for the drpc metadata fixes done earlier
and removes all existing uses of drpcmetadata in the code, replacing them
with grpc metadata calls.

Why this works:
On the client side, drpc client framework code fixes take care of
sending grpc metadata over the wire. On the server side, the
drpc server framework code populates the metadata received over
the wire as grpc context metadata. The drpc server's manager is
instantiated with a newly added option (GRPCMetadataCompactMode)
to configure this behaviour. Thus, both client and server can
simply use grpc metadata calls to send and receive metadata.

Epic: CRDB-48924
Fixes: #157774
Release note: None

suj-krishnan committed:
Merge branch 'master' into fix_drpcmetadata_usage

cthumuluru-crdb reviewed (approved):
LGTM!

suj-krishnan committed:
Merge branch 'master' into fix_drpcmetadata_usage

suj-krishnan committed:
Merge branch 'master' into fix_drpcmetadata_usage

suj-krishnan committed:
Merge branch 'master' into fix_drpcmetadata_usage

suj-krishnan commented:
bors retry
